### PR TITLE
FEATURE: Allow users to set a custom title to sidebar calendar

### PR DESCRIPTION
**Pull request by brrusselburg**
This allows users to change the title of the calendar sidebar according to the category. The example use-case is "Upcoming Releases" in the "Product News" category.

Default:
Screenshot 2024-11-27 at 1 50 45 PM

Custom:
Screenshot 2024-11-27 at 1 51 30 PM

Setting:
Screenshot 2024-11-27 at 1 51 15 PM

Tests will be added into the Right Sidebar Blocks component, as this feature only works in conjunction with specific settings in that component.

### CODE REVIEW

**Comment by renato** [comment]
I think this behavior might be entirely tested within the existing integration test we have here, no? What specifics of the right sidebar blocks we need to cover?

If it's just the category we can mock different categories with the RouterStub for different tests

**Comment by brrusselburg** [comment]
> I think this behavior might be entirely tested within the existing integration test we have here, no? What specifics of the right sidebar blocks we need to cover?
> 
> If it's just the category from the route, we can mock different categories with the RouterStub for different tests

The default title is covered, but I'm not sure I see how a custom path would be 🤔 And yeah, the issue we had the other day was in getting the context of a specific category into this test.

File: assets/javascripts/discourse/components/upcoming-events-list.gjs [hunk 1]
```diff
@@ -68,6 +67,36 @@ export default class UpcomingEventsList extends Component {
     return this.router.currentRoute.attributes?.category?.id;
   }
 
+  get title() {
+    const categorySlug = this.router.currentRoute.attributes?.category?.slug;
+    const titleSetting = this.siteSettings?.map_events_title;
```
Comment (CvX) [inline assets/javascripts/discourse/components/upcoming-events-list.gjs:72]: siteSettings is always present

```suggestion
    const titleSetting = this.siteSettings.map_events_title;
```

File: assets/javascripts/discourse/components/upcoming-events-list.gjs [hunk 2]
```diff
@@ -68,6 +67,36 @@ export default class UpcomingEventsList extends Component {
...
+        (o) => o.category_slug === categorySlug
+      );
+      const custom_title =
+        typeof customTitleLookup === "undefined"
+          ? null
+          : customTitleLookup.custom_title;
+
+      return I18n.t("discourse_post_event.upcoming_events_list.custom_title", {
+        upcoming_events_title: custom_title,
+      });
```
Comment (CvX) [inline assets/javascripts/discourse/components/upcoming-events-list.gjs:94]: golf + case

```suggestion
      const customTitle = customTitleLookup?.custom_title;

      return I18n.t("discourse_post_event.upcoming_events_list.custom_title", {
        upcoming_events_title: customTitle,
      });
```

File: config/locales/client.en.yml [hunk 3]
```diff
@@ -373,6 +373,7 @@ en:
         starts_at: "Starts at"
       upcoming_events_list:
         title: "Upcoming events"
+        custom_title: "%{upcoming_events_title}"
```
Comment (CvX) [inline config/locales/client.en.yml:376]: do we need a i18n string for this?

Comment (brrusselburg) [reply config/locales/client.en.yml:376]: Probably not? I split them to make it easier for me to follow while I worked on the solution.

Comment (CvX) [reply config/locales/client.en.yml:376]: yeah, I'd drop it

File: assets/javascripts/discourse/components/upcoming-events-list.gjs [hunk 4]
```diff
@@ -68,6 +67,31 @@ export default class UpcomingEventsList extends Component {
...
+    );
+
+    if (categories.includes(categorySlug)) {
+      const titleMap = JSON.parse(titleSetting);
+      const customTitleLookup = titleMap.find(
+        (o) => o.category_slug === categorySlug
+      );
+      const custom_title = customTitleLookup?.custom_title;
+
+      return custom_title;
```
Comment (CvX) [inline assets/javascripts/discourse/components/upcoming-events-list.gjs:89]: shorter and with no underscore_case variable:

```suggestion
      return customTitleLookup?.custom_title;
```

File: test/javascripts/integration/components/upcoming-events-list-test.gjs [hunk 5]
```diff
@@ -163,6 +163,41 @@ module("Integration | Component | upcoming-events-list", function (hooks) {
...
+      '[{"category_slug": "announcements", "custom_title": "Upcoming Announcements"}]';
+
+    await render(<template><UpcomingEventsList /></template>);
+    this.appEvents.trigger("page:changed", { url: "/c/announcements" });
+
+    assert.strictEqual(
+      query(".upcoming-events-list__heading").innerText,
+      "Upcoming Announcements",
+      "it sets 'Upcoming Announcements' as the title in 'c/announcements'"
+    );
```
Comment (CvX) [inline test/javascripts/integration/components/upcoming-events-list-test.gjs:181]: qunit-dom assertions are the only recommended way to test text/html/existence/etc going forward

```suggestion
    assert
      .dom(".upcoming-events-list__heading")
      .hasText(
        "Upcoming Announcements",
        "sets 'Upcoming Announcements' as the title in 'c/announcements'"
      );
```

**Comment by brrusselburg** [comment]
> Sorry @brrusselburg ...
> 
> Is there a reason why we're not using text replacement feature here? https://meta.discourse.org/t/customize-text-in-discourse/36092
> 
> http://localhost:4200/admin/customize/site_texts?q=discourse_post_event.upcoming_events_list.title Screenshot 2024-12-03 at 11 18 48 AM

Yep. That was originally being used, but it doesn't account for different titles per route.

File: test/javascripts/integration/components/upcoming-events-list-test.gjs (hunk 5 above)
Comment (brrusselburg) [reply test/javascripts/integration/components/upcoming-events-list-test.gjs:181]: good to know, thank you!

File: assets/javascripts/discourse/components/upcoming-events-list.gjs (hunk 4 above)
Comment (brrusselburg) [reply assets/javascripts/discourse/components/upcoming-events-list.gjs:89]: ope, totally missed that. thank you.

**Review by CvX — APPROVED** [review]

**Pull request merged**